### PR TITLE
Improve nsexec logging

### PR DESCRIPTION
**Pull request by kolyshkin**
~~Draft as it's blocked by #2835~~

In order to make `runc --debug` actually useful for debugging nsexec
bugs, provide information about all the internal operations when in
debug mode.

Before:
```console
# runc --debug run x123
DEBU[0000] nsexec:601 nsexec started                    
DEBU[0000] child process in init()                      
DEBU[0000] logging has already been configured          
DEBU[0000] log pipe has been closed: EOF                
DEBU[0001] process exited                                pid=1472529 status=0
```

After:
```console
# runc --debug run x123
DEBU[0000] nsexec[1472380]: => nsexec container setup   
DEBU[0000] nsexec-0[1472380]: ~> nsexec stage-0         
DEBU[0000] nsexec-0[1472380]: spawn stage-1             
DEBU[0000] nsexec-0[1472380]: -> stage-1 synchronisation loop 
DEBU[0000] nsexec-1[1472384]: ~> nsexec stage-1         
DEBU[0000] nsexec-1[1472384]: unshare remaining namespace (except cgroupns) 
DEBU[0000] nsexec-1[1472384]: spawn stage-2             
DEBU[0000] nsexec-1[1472384]: request stage-0 to forward stage-2 pid (1472385) 
DEBU[0000] nsexec-0[1472380]: stage-1 requested pid to be forwarded 
DEBU[0000] nsexec-0[1472380]: forward stage-1 (1472384) and stage-2 (1472385) pids to runc 
DEBU[0000] nsexec-2[1]: ~> nsexec stage-2               
DEBU[0000] nsexec-1[1472384]: signal completion to stage-0 
DEBU[0000] nsexec-1[1472384]: <~ nsexec stage-1         
DEBU[0000] nsexec-0[1472380]: stage-1 complete          
DEBU[0000] nsexec-0[1472380]: <- stage-1 synchronisation loop 
DEBU[0000] nsexec-0[1472380]: -> stage-2 synchronisation loop 
DEBU[0000] nsexec-0[1472380]: signalling stage-2 to run 
DEBU[0000] nsexec-2[1]: unshare cgroup namespace        
DEBU[0000] nsexec-2[1]: signal completion to stage-0    
DEBU[0000] nsexec-0[1472380]: stage-2 complete          
DEBU[0000] nsexec-0[1472380]: <- stage-2 synchronisation loop 
DEBU[0000] nsexec-0[1472380]: <~ nsexec stage-0         
DEBU[0000] nsexec-2[1]: <= nsexec container setup       
DEBU[0000] nsexec-2[1]: booting up go runtime ...       
DEBU[0000] child process in init()                      
DEBU[0000] init: closing the pipe to signal completion  
DEBU[0000] sending signal to process urgent I/O condition 
DEBU[0000] sending signal to process urgent I/O condition 
DEBU[0001] process exited                                pid=1472385 status=0
```
This is a carry of https://github.com/opencontainers/runc/pull/2460 with a few fixes on top:

 - rebased on top of current master (resolved conflicts due to #2481 etc);
 - fixed C formatting in nsexec.c to satisfy `make cfmt`;
 - added JSON string escaping (fixes debug.bats failures);
 - rebased on top of #2835
 
Closes: #2460 

(this was originally developed in #2487)

### CODE REVIEW

**Comment by thaJeztah** [comment]
This still draft?

**Comment by kolyshkin** [comment]
No longer a draft; @cyphar PTAL (this is mostly your code)

**Comment by kolyshkin** [comment]
Rebased again; PTAL @cyphar

**Comment by kolyshkin** [comment]
centos7 fails
```
=== RUN   TestSystemdFreeze
    utils_test.go:86: exec_test.go:539: unexpected error: unable to freeze
        
--- FAIL: TestSystemdFreeze (0.74s)
```

This is the second time I see it, not sure what's happening but definitely unrelated to this PR.

CI restarted.

**Review by cyphar — DISMISSED** [review]
LGTM. I'll take your word for it that most of this new code is mine. :grimacing:

**Comment by kolyshkin** [comment]
CentOS CI failure is a flake (https://github.com/opencontainers/runc/issues/2760), restarting.

File: libcontainer/nsenter/escape.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	if (j == 0) {
+		// nothing to escape
+		return s;
+	}
+
+	len = i + j + 1;
+	out = malloc(len);
+	if (!out) {
+		exit(1);
```
Comment (mrunalp) [inline libcontainer/nsenter/escape.c:90]: Do we want to return error here instead?

Comment (kolyshkin) [reply libcontainer/nsenter/escape.c:90]: I'd do something like `return strdup("escape_json: out of memory"` but then strdup might fail, too.

Other than that there's no way to return an error, so maybe `return NULL` and handle it in the caller?

Comment (kolyshkin) [reply libcontainer/nsenter/escape.c:90]:
```diff
@@ -87,7 +88,10 @@ char *escape_json_string(char *s)
        len = i + j + 1;
        out = malloc(len);
        if (!out) {
-               exit(1);
+               free(s);
+               // As malloc failed, strdup can fail, too, so in the worst case
+               // scenario NULL will be returned from here.
+               return strdup("escape_json_string: out of memory");
        }
        for (c = s, j = 0; *c != '\0'; c++) {
                switch (*c) {
```

Comment (kolyshkin) [reply libcontainer/nsenter/escape.c]: updated

**Review by mrunalp — APPROVED** [review]

**Comment by kolyshkin** [comment]
@cyphar PTAL; the only change after your review was this: https://github.com/opencontainers/runc/pull/2836#discussion_r612023579

File: libcontainer/nsenter/escape.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		return s;
+	}
+
+	len = i + j + 1;
+	out = malloc(len);
+	if (!out) {
+		free(s);
+		// As malloc failed, strdup can fail, too, so in the worst case
+		// scenario NULL will be returned from here.
+		return strdup("escape_json_string: out of memory");
```
Comment (cyphar) [inline libcontainer/nsenter/escape.c:94]: I feel like it makes more sense to always return NULL (and then in the caller provide the fallback error message), but it doesn't really matter imho.

**Review by cyphar — APPROVED** [review]
LGTM.

**Pull request closed without merging**